### DISCUSSION TITLE
Cant seem to persist session with Expo

### DISCUSSION THREAD

**Discussion by farzd** [supabase/supabase#20155] [Questions]
I've had to update to Expo 49 and Supabase to v2.x for a clients app, a lot of the supabase auth API calls have changed, so has expo linking etc - much better.
Everything is good now, except when I restart the app, I have to login again 
```
const { data: { session }, error } = await supabase.auth.getSession()
```
i'm using this to get the persisted session but its returning null when i restart the app

my setup:
```
import 'react-native-url-polyfill/auto'
import AsyncStorage from '@react-native-async-storage/async-storage'
import { createClient } from '@supabase/supabase-js'

const supabase = createClient(supabaseUrl, supabaseKey, {
  storage: AsyncStorage,
  autoRefreshToken: true,
  persistSession: true,
  detectSessionInUrl: false,
})

export default supabase
```

``` supabase.auth.onAuthStateChange ``` seems to work fine, i can login and get the session 

```
  "@supabase/supabase-js": "^2.39.2",
 "@react-native-async-storage/async-storage": "^1.21.0",
 "react-native-url-polyfill": "^2.0.0",
```

i pretty much have the convention stated here
https://supabase.com/docs/guides/getting-started/tutorials/with-expo-react-native

**Comment by farzd** [comment]
well, i should have paid closer attention to the new API, i was missing 'auth' key in the opts object
```
const supabase = createClient("https://xyzcompany.supabase.co", "public-anon-key", {
  auth: {
    storage: AsyncStorage,
    autoRefreshToken: true,
    persistSession: true,
    detectSessionInUrl: false,
  },
});
```

**Reply by DrSebastianK** [reply]
Hey Farzd? Did this alone resolve the issue? I'm curious, do mobile app session recover, even if user didn't use the app for hours or even days? If yes, could you share a bit more of your current auth setup?

**Reply by farzd** [reply]
yes it persists, you'll need to have some code that checks and redirects appropriately.
For example i'm using expo router and in my root level _layout.js file, i have this 

```
function RootLayoutNav() {
  const router = useRouter();
  const initialize = useAuthStore((state) => state.initialize);

  useEffect(() => {
    initialize();

    const {
      data: { subscription },
    } = supabase.auth.onAuthStateChange((event, session) => {
      useAuthStore.getState().setSession(session);
      if (event === "SIGNED_IN") router.replace("/");
    });

    return () => subscription.unsubscribe();
  }, []);

  return (
    <>
      <Slot />
    </>
  );
}
```
```
import { create } from "zustand";
import supabase from "../supabase";

const useAuthStore = create((set) => ({
  session: null,
  isLoading: true,
  setSession: (session) => set({ session }),
  setIsLoading: (isLoading) => set({ isLoading }),
  signOut: async () => {
    await supabase.auth.signOut();
    set({ session: null });
  },
  initialize: async () => {
    const {
      data: { session },
    } = await supabase.auth.getSession();
    set({ session, isLoading: false });
  },
}));

export default useAuthStore;
```

https://docs.expo.dev/guides/using-supabase/

**Reply by DrSebastianK** [reply]
Interestingly on my end this solution recovers session after about half hour of having the app closed or when restarting app. After few hours of inactivity getSession can't find the session. Anyways I will keep on experimenting with different methods, something will probably solve it. Thanks for your help anyway!

**Reply by farzd** [reply]
i'm pretty sure supabase refreshes your token automatically. Even if i close my app [hard close] and reopen, it hydrates/checks the session and redirects me to my logged in page. I've used this convention in a few apps

**Comment by therealgreasybeefcake** [comment]
I am having this issue too where Supabase sessions won't persist with my Expo app. Did anyone find a solution to this? Would very much be appreciated.

**Comment by DrSebastianK** [comment]
Double check your codebase, there will be code that tries to refresh the
session. Make sure your supabase implementation is as minimal as possible,
as the package handles most tasks in background automatically. Using
getSession or getUser that may refresh your token can interfere with the
background mechanism. If you use supabase authentication session on your
backend wrong, your backend will likely make a new session, interfering
with the client side session, invalidating refresh token. I'm not sure if
everything I wrote is correct, as I had this problem a long time ago. But
the main thing was,to  double check, to have no place which could possibly
generate a new token. There almost certainly is in your case.

therealgreasybeefcake ***@***.***> ezt írta (időpont: 2026.
febr. 20., P, 4:55):

**Reply by therealgreasybeefcake** [reply]
Thank you for your reply. I will take your advice.

**Reply by therealgreasybeefcake** [reply]
@DrSebastianK - following your advice I went back and set up my supabase implementation as minimal as possible and it appears to have fixed the issue. Thanks for your advice.

**Reply by SharmaVrishab** [reply]
@DrSebastianK  i am also following your advice and i found that i was manual refresh on my side that was causing the issue thanks for pointing that out

**Comment by DrSebastianK** [comment]
Glad I could help!2026. febr. 22. dátummal, 22:30 időpontban therealgreasybeefcake ***@***.***> írta:﻿
@DrSebastianK - following your advice I went back and set up my supabase implementation as minimal as possible and it appears to have fixed the issue. Thanks for your advice.
